src/main/kotlin/com/fasterxml/jackson/module/kotlin/KotlinValueInstantiator.kt

```diff
@@ -67,24 +66,20 @@ internal class KotlinValueInstantiator(
             val jsonProp = props[idx]
             val isMissing = !buffer.hasParameter(jsonProp)
 
-            if (isMissing && paramDef.isOptional) {
-                return@forEachIndexed
-            }
-
             val paramType = paramDef.type
-            var paramVal = if (!isMissing || paramDef.isPrimitive() || jsonProp.hasInjectableValueId()) {
+            var paramVal = if (!isMissing || jsonProp.hasInjectableValueId()) {
```

Comment:
Removed the decision because `jsonProp.valueDeserializer?.getAbsentValue(ctxt)` provides the same functionality as this branch.
The following test confirms that it fails if `FAIL_ON_NULL_FOR_PRIMITIVES` is set.
https://github.com/FasterXML/jackson-module-kotlin/pull/738/files#diff-d8cb2007657572bfaed68d85f55829d211695e1f181981499dcb6a3a72516472